File: heaps/heap_max.py
import heapq

# Creating a max heap using the negation trick
max_heap = []

# Adding elements to the max heap
heapq.heappush(max_heap, -5)
heapq.heappush(max_heap, -3)
heapq.heappush(max_heap, -8)
heapq.heappush(max_heap, -1)

# Convert the list into a max heap using heapify
# max_heap = [-5, -3, -8, -1]
# heapq.heapify(max_heap)

print("Max Heap:")
print([-x for x in max_heap])
# print(max_heap) # [-8, -5, -3, -1]

# Popping elements from the max heap
print("Elements popped from max heap:")
print(-heapq.heappop(max_heap))
print(-heapq.heappop(max_heap))
print(-heapq.heappop(max_heap))

print("Max Heap after popping:")
print([-x for x in max_heap])

File: heaps/heap_max.js
const Heap = require('heap');

// Creating a max heap using the negation trick
const maxHeap = new Heap();

// Adding elements to the max heap (multiplying by -1)
maxHeap.push(-5);
maxHeap.push(-3);
maxHeap.push(-8);
maxHeap.push(-1);

console.log("Max Heap (negated values):");
console.log(maxHeap.toArray().map(value => -value)); // Display negated values

// Popping elements from the max heap (multiplying by -1)
console.log("Elements popped from max heap:");
console.log(-maxHeap.pop());
console.log(-maxHeap.pop());
console.log(-maxHeap.pop());

console.log("Max Heap after popping:");
console.log(maxHeap.toArray().map(value => -value)); // Display negated values

